topdown/bytes.go

```diff
@@ -0,0 +1,95 @@
+package topdown
+
+import (
+	"fmt"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+	"math/big"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type multiplier int
+
+const (
+	base = 1024
+
+	none multiplier = 1
+	kb              = base
+	mb              = kb * base
+	gb              = mb * base
+	tb              = gb * base
+)
+
+func numBytesErr(msg string) error {
+	return fmt.Errorf("numbytes error: %s", msg)
+}
+
+var (
+	errNoAmount = numBytesErr("no byte amount provided")
+	errIntConv  = numBytesErr("could not parse amount to integer")
+)
+
+func errUnitNotRecognized(unit string) error {
+	return numBytesErr(fmt.Sprintf("unit %s not recognized", unit))
+}
+
+func builtinNumBytes(a ast.Value) (ast.Value, error) {
+	var m multiplier
+
+	numPattern := `([0-9]+)`
+	strPattern := `([a-z]+)`
+
+	raw, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	s := strings.ToLower(raw.String())
+
+	numRe, err := regexp.Compile(numPattern)
+	if err != nil {
+		return nil, err
+	}
+
+	strRe, err := regexp.Compile(strPattern)
+	if err != nil {
+		return nil, err
+	}
+
+	numStr := numRe.FindString(s)
+	unit := strRe.FindString(s)
+
+	if numStr == "" {
+		return nil, errNoAmount
+	}
+
+	num, err := strconv.Atoi(numStr)
+	if err != nil {
+		return nil, errIntConv
+	}
+
+	switch unit {
+	case "":
+		m = none
+	case "k", "kb", "ki":
+		m = kb
+	case "m", "mb", "mi":
+		m = mb
+	case "g", "gb", "gi":
+		m = gb
+	case "t", "tb", "ti":
```

Comment:
@patrick-east Okay, I see now that this is less straightforward than I realized: https://ux.stackexchange.com/questions/13815/files-size-units-kib-vs-kb-vs-kb. I'm certainly 100% willing to bow to common usage and the principle of least surprise.

Reply:
I guess in the interest in picking something other people do we could lean on what kubernetes does and use the strict definitions, defaulting the single character (`k`, `m`, etc) to the SI one.

This means that we'd have something like:

```
parse_num_bytes("1KB") == 1000
parse_num_bytes("1KiB") == 1024
```

This should be pretty easy to describe in the docs with an example what what is above, and mentioning it is strict on binary vs decimal values.

It also has the advantage of being explicit so that in the case where it maybe isn't what a user expected, it is the arguably more correct thing.